### PR TITLE
[util] Add method http_get

### PR DESCRIPTION
**Pull request by dagurval**

### CODE REVIEW

**Comment by dagurval** [comment]
I need this for fetching runtime statistics from electrum server. Pushing as a separate pull request to check if all the new boost stuff works on travis.

**Comment by sickpig** [comment]
@dagurval 

```
../../src/utilhttp.cpp:4:10: fatal error: 'boost/beast/core.hpp' file not found
#include <boost/beast/core.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~
```

**Comment by dagurval** [comment]
Looks like this requires boost ~1.65~1.66, while BU is on boost 1.64. Can we upgrade? :-)

**Comment by sickpig** [comment]
@dagurval

we could upgrade the dependencies to use boost 1.65 rather than 1.64.

Even bionic come with 1.65. If we decide to upgrade and ask boost >= 1.65 as requirement we should provide some guidance to the people who are using older but still supported distro and are not using gitian binaries or ubuntu ppa repo. 

 
What other people think?

**Comment by dagurval** [comment]
Re-written to use libevent instead of boost. This is worse, but it doesn't require newer boost functionality.

**Comment by sickpig** [comment]
@dagurval thanks for doing it!

I wonder if we could move all out util*.{cpp,h} in separate dir, maybe `src/util`?

This something for another PR thou.

File: src/test/utilhttp_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+#include "utilhttp.h"
+#include "test/test_bitcoin.h"
+#include <boost/test/unit_test.hpp>
+#include <iostream>
+
+BOOST_FIXTURE_TEST_SUITE(utilhttp_tests, BasicTestingSetup)
+
+BOOST_AUTO_TEST_CASE(http_get_test)
+{
+    // std::cout << http_get("127.0.0.1", 8000, "/").str();
```
Comment (gandrewstone) [inline src/test/utilhttp_tests.cpp]: how can we test?  maybe use libevent to make a simple server in the test itself?  https://kukuruku.co/post/lightweight-http-server-in-less-than-40-lines-on-libevent-and-c-11/

File: src/utilhttp.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    {
+        args->http_code = CB_UNKNOWN_ERROR;
+        return;
+    }
+    args->http_code = evhttp_request_get_response_code(req);
+    char buffer[1024];
+    int s = 0;
+    do
+    {
+        assert(static_cast<size_t>(s) <= sizeof buffer);
```
Comment (gandrewstone) [inline src/utilhttp.cpp]: I'd like to see a comment describing why this assert should hold.  Something like "evbuffer_remove should never return a value greater than the buffer we send to it."  

Also, please use a DbgAssert and returning CB_UNKNOWN_ERROR if the assertion fails.  I don't think that it makes sense to stop the client in release mode if there's a problem with this http interface.

File: src/utilhttp.cpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    } while (s > 0);
+};
+
+
+std::stringstream http_get(const std::string &host, const int port, const std::string &target)
+{
+    EventRAII event;
+    std::stringstream res;
+    int http_code = CB_UNKNOWN_ERROR;
+    CallbackArgs cbargs{http_code, res};
```
Comment (gandrewstone) [inline src/utilhttp.cpp:88]: I think that the lifecycle of cbargs needs to extend beyond this function.  It becomes the pointer that passed into response_cb when called, right?

Comment (gandrewstone) [reply src/utilhttp.cpp:88]: EDIT, never mind, looks like you dispatch below, turning this into a synchronous request

File: src/test/utilhttp_tests.cpp (hunk 1 above)
Comment (dagurval) [reply src/test/utilhttp_tests.cpp]: Yes, but do we want the unit tests to start opening sockets?

This is also indirectly tested in the elecrum_basics test in #1633

File: src/utilhttp.cpp [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    {
+        throw std::runtime_error("evhttp_add_header failed");
+    }
+    if (evhttp_add_header(output_headers, "Connection", "close") != 0)
+    {
+        throw std::runtime_error("evhttp_add_header failed");
+    }
+
+    // Ferform request
+    int rc = evhttp_make_request(event.http_conn, event.req, EVHTTP_REQ_GET, target.c_str());
```
Comment (ptschip) [inline src/utilhttp.cpp:124]: nit: typo?  

-// Ferform request
+// Perform request

**Comment by gandrewstone** [comment]
DbgAssert(static_cast<size_t>(s) <= sizeof buffer, {return });

has an error.  just use:

        DbgAssert(static_cast<size_t>(s) <= sizeof buffer, return);

File: src/test/utilhttp_tests.cpp (hunk 1 above)
Comment (dagurval) [reply src/test/utilhttp_tests.cpp]: Alright, let's do this! A custom http server implemented to test a simple http getter.

**Pull request merged**